Add tutorial section for documentation (#6213)
holoviz/panel: doc/tutorials/basic/build_chatbot.md

# Build a Chat Bot

In this tutorial we will build a streaming *chat bot*. We will first use the *high-level* [`ChatInterface`](../../reference/chat/ChatInterface.ipynb) to build a basic chat bot. Then will will add streaming and finally we will make it scale to many users.

:::{admonition} Note
When I ask you to *run the code* in the sections below, you may either execute the code directly in the Panel docs via the green *run* button, in a cell in a notebook or in a file `app.py` that is served with `panel serve app.py --autoreload`.
:::

## Build a Basic Chat Bot

Run the code below:

```{pyodide}
import panel as pn
from time import sleep

pn.extension()

> WIND_TURBINE = "A *wind turbine* is a device that converts the kinetic energy of wind into electrical energy."
> DONT_KNOW = "Sorry. I don't know."
> 
> def get_response(contents, user, instance):
>     if "turbine" in contents.lower():
>         response = WIND_TURBINE
>     else:
>         response = DONT_KNOW

Reviewer: Do we need to separate the response as constants above? Feels like that requires a tad more mental gymnastics to reference it above. Just a thought; I have no strong feelings for either way.
```suggestion
def get_response(contents, user, instance):
    if "turbine" in contents.lower():
        response = "A wind turbine converts wind energy into electricity."
    else:
        response = "Sorry, I don't know."
```

    sleep(1)
    return response

chat_bot = pn.chat.ChatInterface(callback=get_response, max_height=500)
chat_bot.send("Ask me what a wind turbine is", user="Assistant", respond=False)
chat_bot.servable()
```

Try entering `What is a wind turbine?` in the *text input* and click *Send*.

:::{admonition}
The `callback` function `get_response` will receive `"What is a wind turbine?"` in the `contents` argument. Since `contents` contains the word *turbine*, the chat bot will return `WIND_TURBINE` as the `response`.
:::

## Add Streaming

> We will now make the chat bot *stream* its response just like chatGPT does.

Reviewer: ```suggestion
We will now make the chat bot *stream* its response just like ChatGPT does.
```

Run the code below:

```{pyodide}
import panel as pn
from time import sleep

pn.extension()

WIND_TURBINE = "A *wind turbine* is a device that converts the kinetic energy of wind into electrical energy."
DONT_KNOW = "Sorry. I don't know."

def get_response(contents, user, instance):
    if "turbine" in contents.lower():

>         response = WIND_TURBINE.split(" ")
>     else:
>         response = DONT_KNOW.split(" ")

Reviewer: ```suggestion
        response = WIND_TURBINE
    else:
        response = DONT_KNOW
```

Author: The `.split` is needed to stream tokens back. Otherwise you stream characters back. The first is what happens in the real world. But the .split`. Adds more complexity.

Not sure what to do here?

>     

Reviewer: ```suggestion
    message = ""
```

>     for index in range(len(response)):

Reviewer: ```suggestion
    for char in response:
```

>         yield " ".join(response[:index+1])

Reviewer: Yield here doesn't actually work because it's not an async generator I think.

```suggestion
        instance.stream(char, user="Assistant")
```

Author: I don't understand? I have tested that it works.

        sleep(0.10)

chat_bot = pn.chat.ChatInterface(callback=get_response, max_height=500)
chat_bot.send("Ask me anything!", user="Assistant", respond=False)
chat_bot.servable()
```

Try entering `What is a wind turbine?` in the *text input* and click *Send*.

:::{admonition} Note
The chat app is now *streaming* because we `yield` the partial `response` instead of `return`ing the full `response`.
:::

## Make it scale

:::{admonition}
If your chat bot is successful and many users start using it, you will find that they are *blocking* each other. While one user is getting her response, another user will be waiting on his response.

> You can avoid blocking in many ways. Here we will use an `async` `callback` function.

Reviewer: Should note sleep is now imported from asyncio. Also not entirely sure if this should be under "basic".

Author: The reason why I put it in, is if we don't use `async` there is a big risk the application ends up not performing in the real world. But it also adds more complexity. Maybe just mention that more advanced topics like `async` are referred to the intermediate tutorial?

    if "turbine" in contents.lower():
        response = "A wind turbine converts wind energy into electricity."
    else:
        response = "Sorry, I don't know."
    sleep(1)
    return response

chat_bot = pn.chat.ChatInterface(callback=get_response, max_height=500)
chat_bot.send("Ask me what a wind turbine is", user="Assistant", respond=False)
chat_bot.servable()
```

Try entering `What is a wind turbine?` in the *text input* and click *Send*.

:::{note}
The `callback` function `get_response` will receive `"What is a wind turbine?"` in the `contents` argument. Since `contents` contains the word *turbine*, the chat bot will return `"A wind turbine converts wind energy into electricity."` as the `response`.
:::

## Add Streaming

We will now make the chat bot *stream* its response just like ChatGPT does.

Run the code below:

```{pyodide}
import panel as pn
from time import sleep

pn.extension()

def get_response(contents, user, instance):
    if "turbine" in contents.lower():
        response = "A wind turbine converts wind energy into electricity."
    else:
        response = "Sorry, I don't know."
    for index in range(len(response)):
        yield response[0:index+1]

>         sleep(0.03)

Author: Yes. I've added a comment to the code: `# to simulate slowish response`.

chat_bot = pn.chat.ChatInterface(callback=get_response, max_height=500)
chat_bot.send("Ask me anything!", user="Assistant", respond=False)
chat_bot.servable()
```

Try entering `What is a wind turbine?` in the *text input* and click *Send*.

:::{note}
The chat app is now *streaming* because we `yield` the partial `response` instead of `return`ing the full `response`.
:::

:::{note}

> To make the Streaming Chat Bot scale to many users you should be using `async`. You will learn more about scaling applications and `async` in the [Intermediate Tutorials](../intermediate/index.md).

Author: If you want to support many users `sync` is an antipattern. Just as using `.show` instead of `panel serve`. But single user/ few user use cases also have their place. That is where a lot of people start. Or its what they use for exploration.

I think its ok. @ahuang11 . Any recommendations on this comment?

> To make the Streaming Chat Bot scale to many users you should be using `async`. You will learn more about scaling applications and `async` in the [Intermediate Tutorials](../intermediate/index.md).

Reviewer 2: This is why I believe building a chat app would be worth its own tutorial, to show and explain this sort of thing.

> To make the Streaming Chat Bot scale to many users you should be using `async`. You will learn more about scaling applications and `async` in the [Intermediate Tutorials](../intermediate/index.md).

Author: But this is that tutorial. If it needs more explanation or detail any one can make a PR to make it better.

> To make the Streaming Chat Bot scale to many users you should be using `async`. You will learn more about scaling applications and `async` in the [Intermediate Tutorials](../intermediate/index.md).

Reviewer 2: What I was trying to say is that I believe there should only be one "build a chat app tutorial" and not a basic one, an intermediate one, an advanced one, etc. We have to make decisions, that's hard, but that's better for the user in the end.

See it like the problem we have with APIs. We have too many of them that do the same-ish thing. Let's not apply the same scheme for the tutorials, users will have the How-Tos, Reference and Explanations to learn more and find out about the rest of things you can do.

> To make the Streaming Chat Bot scale to many users you should be using `async`. You will learn more about scaling applications and `async` in the [Intermediate Tutorials](../intermediate/index.md).

Reviewer 3: I agree that this is that tutorial, whether it lives in basic or gets it's own entirely separate top-level section is up in the air I think but we did already decide to separate this out from the main flow of the basic tutorial. In the long run I do aim to have multiple sections for the chat app tutorial but for now this seems totally fine.
